.changeset/happy-zoos-shout.md

```diff
@@ -0,0 +1,9 @@
+---
+'@cloudflare/next-on-pages': patch
+---
+
+make sure that server actions can work in `wrangler pages dev`
+
+fix `wrangler pages dev` breaking server actions because of a misalignment between headers
```

Comment:
Question: If this is only intended to be for wrangler pages dev, would it be problematic or introduce a security concern if someone deploys the build?

Reply:
It shouldn't introduce any problem/security concern, if it would then I don't think the change would be worth it and we could instead ask people to set an appropriate `allowedOrigins` for during local development (it could rely on an `.env.local` file making it, although still inconvenient, not terrible).

It would be a bit cumbersome so I would prefer to try to avoid that if we could 😓

Regarding why I don't think it's introducing a risk, the fact is that by the code below you can see that I am updating a pretty specific line and I add a regex check that strictly changes logic only on hosts like: `localhost:xxxx` so I feel like it's not introducing any risks.

(then we also update a `console.warn` but even if the changes break that I don't think that's not the end of the world)

More than introducing security concerns I am a bit afraid that this is a bit brittle and can easily break... hopefully not but I guess we can only be sure by actually seeing how it performs in the wild 😅

Reply:
But if you see a security concern or something that looks problematic to you please let me know 🙏 
(I think this could be quite important for local development but for sure, better to break local dev for users instead of breaking their production applications 😅)

Reply:
Hmm, I'm really not sure I like the idea of doing a modification to something that has a high chance of being deployed to production, when the modification is only supposed to be presented in one single specific dev environment...

I really think that we need a way to determine if the user is using wrangler pages dev if we want to do this kind of hack. These kind of changes/hacks we've done in the past have always been intended to be shipped to production, whereas this change is only intended to be used in wrangler pages dev.

And since the Next.js PR for wildcards got merged, then surely it would be better to leave this to the user if they want to do this in their next.config.js instead. It might take a week for a new Next.js release, but to me that is a better choice than making modifications to their CSRF protections that would be shipped to prod when it shouldn't be.

Reply:
As I mentioned `allowedOrigins` even when the wildcard change is released, is going to be pretty tedious and error prone in my opinion (maybe having a local address in the as an allowed origin in production is not going to be an issue, so people could just add it there and leave it... I am not really sure). Ideally I would really not want developers to have to deal with it themselves.

Regarding detecting if the build output is going to be used for `wrangler pages dev` I think that's not really a great idea based on how the adapter works. Since it just builds the worker for you, I don't think it should have any knowledge on how you are going to use that worker. Actually, I can imagine people building their worker with next-on-pages, previewing it with `wrangler pages dev` and then deploying it with `wrangler pages deploy`, so the same exact worker should work in both dev and production (and it should be the same worker since people are trying to validate it before the deployment, if we produced different workers then people could not actually accurately "preview" the actual production worker before deploying it).

I think it is very clear that modifying production code for the sake of local development is not ideal in the slightest, I just think that the changes here are so negligible (both in terms of performance and security) that it's worth it in this case given the DX gain.

Reply:
> I just think that the changes here are so negligible (both in terms of performance and security)

But this is a limitation that Next.js specifically added for production builds to help prevent CSRF attacks, so I would really have to disagree that this is negligible. They wouldn't go through the effort to add this additional protection if there wasn't any value in it.

If we are modifying it to allow localhost:{num} to be bypass this protection in any deployment, then IMO that is a very slippery slope and could end up being consequential for some of our users.

> Regarding detecting if the build output is going to be used for wrangler pages dev I think that's not really a great idea based on how the adapter works

I meant through something like an env variable existing from wrangler to dictate that this was in pages dev as a *gate* for when this hack works in their code. something like `process.env.WRANGLER_DEV && /.../.test(...) ? null : origin`, but i dont know if that is currently possible with wrangler. this would then give the desired behaviour in local dev, and the correct behaviour/protections in deployments.

Reply:
> I meant through something like an env variable existing from wrangler to dictate that this was in pages dev as a _gate_ for when this hack works in their code. something like `process.env.WRANGLER_DEV && /.../.test(...) ? null : origin`, but i dont know if that is currently possible with wrangler. this would then give the desired behaviour in local dev, and the correct behaviour/protections in deployments.

ah ok, yeah that sounds like an interesting idea, I misunderstood what you were suggesting 🙂

Reply:
anyways thanks a lot for the insights 🙂

but as we were discussing this with @petebacondarwin, it sounds like maybe things can be updated in wrangler so to have it set a proper `x-forwarded-host` when passing requests from the proxy worker to the actual worker without any hacky changes needed in next-on-pages

so I'm going to close the PR for now and see if things can be changes in wrangler instead, if not we can revisit things here later 🙂

Reply:
ah nice, that actually sounds like a pretty good middleground!